reference/best_practices/reusable_bundles.rst

```diff
@@ -1,4 +1,173 @@
+
+Try your bundle with other catalogs. Don't forget use cases with many locales, many channels and various catalog volumes.
+You can use our `data generator bundle`_ to generate a catalog with random data.
+
+.. _Akeneo PIM Community Edition: https://github.com/akeneo/pim-community-standard
+.. _Enterprise Edition: https://www.akeneo.com/enterprise-edition/
+.. _data generator bundle: https://github.com/akeneo-labs/DataGeneratorBundle
+
+
+Maintainability
+---------------
+
+Keep in mind the `Single Responsibility`_ from the five `SOLID principles`_ of Object Oriented Programming.
+It seems simple but we often forget to explode your code in multiple classes to make it more readable and testable.
+
+You can use `PMD`_ to help make your code simple, flexible and easily testable.
+
+
+In Akeneo PIM, we use `PhpSpec`_ as unit testing tool and `Behat`_ for behavior-oriented tests.
+
+Use a Continuous Integration (CI) tool to deploy your application, test your code and improve its quality.
+On our open-source projects, we use free CI tools `Travis`_ for unit tests and `Scrutinizer`_ for static analysis.
+You can find some examples on our `Akeneo-Labs`_ projects (`scrutinizer.yml`_ and `travis.yml`_).
+
+.. _Single Responsibility: https://en.wikipedia.org/wiki/Single_responsibility_principle
+.. _SOLID principles: https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)
+.. _PMD: https://phpmd.org/
+.. _PhpSpec: http://phpspec.readthedocs.org/
+.. _Behat: http://docs.behat.org/
+.. _Travis: https://travis-ci.org/
+.. _Scrutinizer: https://scrutinizer-ci.com/
+.. _Akeneo-Labs: https://github.com/akeneo-labs
+.. _scrutinizer.yml: https://github.com/akeneo-labs/CustomEntityBundle/blob/master/.scrutinizer.yml
+.. _travis.yml: https://github.com/akeneo-labs/CustomEntityBundle/blob/master/.travis.yml
+
+
+Another important point about maintainability is to try to rely on interfaces not on concrete classes.
+An interface is a contract you will inject in a method. It will ease you the migration on future Akeneo PIM versions.
+
+Don't forget you can use composition instead of inheritance.
+Decorate a class will allow you to benefit of inheritance avoiding migration problems.
+Indeed, class constructors can changed from a version to another. On the other side, an interface do not change.
+
+To give to integrator a clean way to rely on your code, you can alse create interfaces.
+
+
+Extensibility
+-------------
+
+Respect Akeneo PIM dependencies
+"""""""""""""""""""""""""""""""
+You **MUST** follow Akeneo PIM dependencies at tall cost. Don't try to change version of existing ones.
+You may add new dependencies needed for your customisations.
+
+
+Use registries
+""""""""""""""
+Registries are also a good extension point when you have to manage with many cases (different attribute types, entities, etc.).
+It consists in creating a registry class and then tag services which will be add to this registry.
```

Comment:
Registries are used to gather related classes (i.e: filters, sorters). You can add new classes to a registry simply by tagging it. 
In general you only need to use existing tags.